tools/cpp/modules_tools/common/common.cc

```diff
@@ -0,0 +1,139 @@
+    return dep;
+  }
+  auto rule_data = rules[0];
+  if (!rule_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto rule = rule_data.as_object();
+  auto provides_data = rule["provides"];
+  if (!provides_data.is_array()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto provides = provides_data.as_array();
+  if (provides.size() > 1) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  if (provides.size() == 1) {
+    auto provide_data = provides[0];
+    if (!provide_data.is_object()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    auto name_data = provide_data.as_object().at("logical-name");
+    if (!name_data.is_string()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    dep.gen_bmi = true;
+    dep.name = name_data.as_string();
+  }
+  auto requires_data = rule["requires"];
+  if (!requires_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  for (const auto &item_data : requires_data.as_array()) {
+
+    if (!item_data.is_object()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    auto name_data = item_data.as_object().at("logical-name");
+    if (!name_data.is_string()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    dep.require_list.push_back(name_data.as_string());
+  }
+  return dep;
+}
+Cpp20ModulesInfo parse_info(std::istream &info_stream) {
+  std::string info_string((std::istreambuf_iterator<char>(info_stream)),
+                          std::istreambuf_iterator<char>());
+  JsonValue data = parse_json(info_string);
+  if (!data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto modules_data = data.as_object().at("modules");
+  if (!modules_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto usages_data = data.as_object().at("usages");
+  if (!usages_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  Cpp20ModulesInfo info;
+  for (const auto &item_data : modules_data.as_object()) {
+    auto name = item_data.first;
+    auto bmi_data = item_data.second;
+    if (!bmi_data.is_string()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    info.modules[name] = bmi_data.as_string();
+  }
+  for (const auto &item_data : usages_data.as_object()) {
+    auto name = item_data.first;
+    auto require_list_data = item_data.second;
+    if (!require_list_data.is_array()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    std::vector<std::string> require_list;
+    for (const auto &require_item_data : require_list_data.as_array()) {
+      if (!require_item_data.is_string()) {
+        std::cerr << "bad ddi" << std::endl;
+        std::exit(1);
+      }
+      require_list.push_back(require_item_data.as_string());
```

Comment:
Would `emplace_back` allow a move rather than a copy?

Reply:
emm, it seems not.

using `emplace_back` would not eliminate the overhead of copying. This is because `require_item_data` still holds the data, and `as_string` returns a `const std::string&`.